### PR TITLE
add flow-informed tuning guidance

### PR DESCRIPTION
**Pull request by yaauie**
## Release notes

[rn: skip]

## What does this PR do?

Improves Tuning docs to use flow-metrics to inform intuition about how and why to tune what.

## Why is it important/What is the impact to the user?

Increases visibility to Flow metrics so that those seeking to tune LS pipeline performance can find a useful tool.

## Checklist

- ~~[ ] My code follows the style guidelines of this project~~
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- [x] I have made corresponding changes to the documentation
- ~~[ ] I have made corresponding change to the default configuration files (and/or docker env variables)~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~

## How to test this PR locally

Wait for the docs gen link to be generated :)

### CODE REVIEW

File: docs/static/performance-checklist.asciidoc [hunk 1]
```diff
@@ -49,10 +49,43 @@ sure-fire way to create a confusing situation.
...
+Depending on the input plugin being used, back-pressure can either propagate upstream or result in the risk of data-loss.
+
+A pipeline's `queue_backpressure` flow metric reflects how much time the inputs are spending attempting to push events into the queue.
+The metric isn't precisely comparable across pipelines, but instead allows you to compare a single pipeline's current behaviour to _itself_ over time.
+When this metric is growing, we generally need to look _downstream_ at the pipeline's filters and outputs to see if they are using resources effectively, have sufficient resources allocated, or are experiencing back-pressure of their own.
+
+NOTE: A persisted queue offers durability guarantees and can absorb back-pressure for longer than the default in-memory queue, but once it is full it too propagates back-pressure.
+      The `queue_persisted_growth_events` flow metric is useful measure of how much back-pressure is being actively absorbed by the persisted queue, and should trend toward zero (or less) over the pipeline's lifetime.
 
 [[tuning-logstash]]
```
Comment (yaauie) [inline docs/static/performance-checklist.asciidoc:90]: It is not clear to me why this section-break results in a separate page. I had hoped that the bits above (the why) would be more closely-tied to the bits below (the how).

Is it just the heading level, or is there something else?

Comment (karenzone) [inline docs/static/performance-checklist.asciidoc:90]:
```suggestion
      The `queue_persisted_growth_events` flow metric is useful measure of how much back-pressure is being actively absorbed by the persisted queue, and should trend toward zero (or less) over the pipeline's lifetime.

[float]
[[tuning-logstash]]
```

Comment (karenzone) [reply docs/static/performance-checklist.asciidoc:90]: Is the presentation you want similar to this?  If so, the `[float]` tag is the magic you're looking for. 

Screen Shot 2024-06-27 at 4 58 59 PM

**Review by karenzone — COMMENTED** [review]
The Elastic standard is to use sentence capitalization for heads, and we haven't touched these docs in a while. So "Tuning and Profiling Logstash Performance" should be  "Tuning and profiling Logstash performance," for example.  Would you consider updating the headings while you're in there?  If you don't have time, I can fix them up in a follow-up PR.

File: docs/static/performance-checklist.asciidoc (hunk 1 above)
Comment (yaauie) [reply docs/static/performance-checklist.asciidoc:90]: After a separate conversation in slack, I think I actually need to move the content to be _inside_ of the next heading, to set context before that page gets into the specifics.

Thank you! Knowing that `[float]` means keep-with-previous helps a lot.

File: docs/static/performance-checklist.asciidoc [hunk 2]
```diff
@@ -49,59 +50,132 @@ sure-fire way to create a confusing situation.
...
+When tuning Logstash you may have to adjust the heap size.
+You can use the https://visualvm.github.io/[VisualVM] tool to profile the heap.
+The *Monitor* pane in particular is useful for checking whether your heap allocation is sufficient for the current workload.
+The screenshots below show sample *Monitor* panes.
+The first pane examines a Logstash instance configured with too many inflight events.
+The second pane examines a Logstash instance configured with an appropriate amount of inflight events.
+Note that the specific batch sizes used here are most likely not applicable to your specific workload, as the memory demands of Logstash vary in large part based on the type of messages you are sending.
```
Comment (jsvd) [inline docs/static/performance-checklist.asciidoc:160]: not worth changing now, but typically I prefer to put style changes in a separate PR to facilitate reviewing, especially in documentation where github isn't great at showing diffs.

File: docs/static/performance-checklist.asciidoc [hunk 3]
```diff
@@ -49,13 +49,49 @@ sure-fire way to create a confusing situation.
...
+Using this information, we can gain intuition about how to tune the pipeline's settings to add resources, or how to find and eliminate wasteful computation.
+
+In general, plugins fit into one of two categories:
+
+- *CPU-bound*: plugins that perform computation on the contents of events _without_ the use of the network or disk IO tend to benefit from incrementally increasing `pipeline.workers` as long as the process has available CPU; once CPU is exhausted additional concurrency can result in _lower_ throughput as the pipeline workers contend for resources.
+- *IO-bound*: plugins that use the network to either enrich events or transmit events tend to benefit from incrementally increasing `pipeline.workers` and/or tuning the pipeline batching as long as those external services are not exerting back-pressure (even if Logstash is using nearly all of its available CPU).
+
+When a pipeline's `worker_utilization` is _below_ 100, its pipeline workers are spending their time waiting for input.
```
[the remark this answers is not in the document]
Comment (yaauie) [reply docs/static/performance-checklist.asciidoc:77]:
```suggestion
The further a pipeline's `worker_utilization` is from 100, the more time its workers are spending waiting for events to arrive in the queue.
```

File: docs/static/performance-checklist.asciidoc [hunk 4]
```diff
@@ -49,59 +50,132 @@ sure-fire way to create a confusing situation.
...
+A pipeline's `queue_backpressure` flow metric reflects how much time the inputs are spending attempting to push events into the queue.
+The metric isn't precisely comparable across pipelines, but instead allows you to compare a single pipeline's current behaviour to _itself_ over time.
+When this metric is growing, we generally need to look _downstream_ at the pipeline's filters and outputs to see if they are using resources effectively, have sufficient resources allocated, or are experiencing back-pressure of their own.
+
+NOTE: A persisted queue offers durability guarantees and can absorb back-pressure for longer than the default in-memory queue, but once it is full it too propagates back-pressure.
+The `queue_persisted_growth_events` flow metric is useful measure of how much back-pressure is being actively absorbed by the persisted queue, and should trend toward zero (or less) over the pipeline's lifetime.
```
[the remark this answers is not in the document]
Comment (yaauie) [reply docs/static/performance-checklist.asciidoc:92]:
```suggestion
The `queue_persisted_growth_events` flow metric is useful measure of how much back-pressure is being actively absorbed by the persisted queue, and should trend toward zero (or less) over the pipeline's lifetime.
Negative numbers indicate that the queue is _shrinking_, and that the workers are catching up on lag that had previously developed.
```

File: docs/static/performance-checklist.asciidoc [hunk 5]
```diff
@@ -49,59 +50,132 @@ sure-fire way to create a confusing situation.
...
+
+Make sure you've read the <<performance-troubleshooting>> before modifying these options.
 
-* The `pipeline.batch.delay` setting rarely needs to be tuned. This setting adjusts the latency of the Logstash pipeline. Pipeline batch delay is the maximum amount of time in milliseconds that Logstash waits for new messages after receiving an event in the current pipeline worker thread. After this time elapses, Logstash begins to execute filters and outputs.The maximum time that Logstash waits between receiving an event and processing that event in a filter is the product of the `pipeline.batch.delay` and  `pipeline.batch.size` settings.
+* The `pipeline.workers` setting determines how many threads to run for filter and output processing.
+  If you find that events are backing up, or that the CPU is not saturated, consider increasing the value of this parameter to make better use of available processing power.
+  Good results can even be found increasing this number past the number of available processors as these threads may spend significant time in an I/O wait state when writing to external systems.
+  Legal values for this parameter are positive integers.
```
[the remark this answers is not in the document]
Comment (yaauie) [reply docs/static/performance-checklist.asciidoc:109]:
```suggestion
```

I agree -- it was here because of the net-no-change refactor to support the one-sentence-per-line convention, but can be ✂️ here.

File: docs/static/performance-checklist.asciidoc [hunk 6]
```diff
@@ -49,59 +50,132 @@ sure-fire way to create a confusing situation.
...
+
+* The `pipeline.batch.size` setting defines the maximum number of events an individual worker thread collects before attempting to execute filters and outputs.
+  Larger batch sizes are generally more efficient, but increase memory overhead.
+  Some hardware configurations require you to increase JVM heap space in the `jvm.options` config file to avoid performance degradation (See <<config-setting-files>> for more info).
+  Values in excess of the optimum range cause performance degradation due to frequent garbage collection or JVM crashes related to out-of-memory exceptions.
```
[the remark this answers is not in the document]
Comment (yaauie) [reply docs/static/performance-checklist.asciidoc:114]: They really do. I'll snip.

```suggestion
```

**Review by jsvd — APPROVED** [review]
LGTM

**Comment by yaauie** [comment]
@logstashmachine backport 8.15

**Comment by yaauie** [comment]
@logstashmachine backport 8.14
